### PR TITLE
Multivariate effect sizes and multiple regression charts

### PR DESCRIPTION
**Pull request by TomiSzucs**

### CODE REVIEW

File: cogstat/cogstat_chart.py [hunk 1]
```diff
@@ -725,6 +725,107 @@ def create_variable_pair_chart(data, meas_lev, x, y, result=None, raw_data=False
...
+    Returns
+    -------
+    matplotlib chart
+        A matrix plot of the variables optionally containing the raw data.
+    """
+
+    if meas_lev == "int":
+        if type == "collinearity":
+            data = data[x]
+        if type in "raw":
```
Comment (krajcsi) [inline cogstat/cogstat_chart.py:750]:
```suggestion
        if type == "raw":
```

File: cogstat/cogstat_chart.py [hunk 2]
```diff
@@ -725,6 +725,107 @@ def create_variable_pair_chart(data, meas_lev, x, y, result=None, raw_data=False
...
+            fig.suptitle(_plt("Scatterplot matrix of variables"))
+            for i in range(0, len(data.columns)):
+                ax[i, 0].set_ylabel(data.columns[i])
+                ax[len(data.columns) - 1, i].set_xlabel(data.columns[i])
+                for j in range(0, len(data.columns)):
+                    if i == j:
+                        ax[i, j].hist(data.iloc[:, i])
+                    else:
+                        ax[i, j].scatter(data.iloc[:, i], data.iloc[:, j])
+        elif type is "collinearity":
```
Comment (krajcsi) [inline cogstat/cogstat_chart.py:761]:
```suggestion
        elif type == "collinearity":
```

File: cogstat/cogstat_chart.py [hunk 3]
```diff
@@ -725,6 +725,107 @@ def create_variable_pair_chart(data, meas_lev, x, y, result=None, raw_data=False
...
+    """Draw a chart relating more than two variables displaying raw data
+
+    Parameters
+    ----------
+    data : pandas dataframe
+    meas_lev : {'int', 'ord', 'nom', 'unk'}
+        Measurement level of the variables
+    x : str
+        Name of the x variable.
+    type : {'raw', 'collinearity'}
```
Comment (krajcsi) [inline cogstat/cogstat_chart.py:738]: if there is no shared code between the two modes, then probably it is better to split the function into two separate functions

File: cogstat/cogstat_chart.py [hunk 4]
```diff
@@ -725,6 +725,107 @@ def create_variable_pair_chart(data, meas_lev, x, y, result=None, raw_data=False
...
+                ncols = 1
+            elif len(x) >= 2:
+                ncols = 2
+            import math
+            nrows = math.ceil(len(x)/2)
+            fig = plt.figure(tight_layout=True)
+            fig.suptitle(_plt("Scatterplot matrix of explanatory variables"))
+            x_done = []
+            ind = 1
+            for index_1, x_i in enumerate(x):
```
Comment (krajcsi) [inline cogstat/cogstat_chart.py:772]: then x is not str but a list of str? plase, modify the docstring (and rename the x related variables here too)

File: cogstat/cogstat_chart.py [hunk 5]
```diff
@@ -725,6 +725,107 @@ def create_variable_pair_chart(data, meas_lev, x, y, result=None, raw_data=False
...
+def part_regress_plots(y, x, data):
+    """Draw a matrix of partial regression plots.
+
+    Parameters
+    ----------
+    x : str
+        Name of the x variable.
+    y : str
+        Name of the y variable.
+    data : pandas dataframe
```
Comment (krajcsi) [inline cogstat/cogstat_chart.py:796]: please, move this parameter to the first position

File: cogstat/cogstat_chart.py [hunk 6]
```diff
@@ -725,6 +725,107 @@ def create_variable_pair_chart(data, meas_lev, x, y, result=None, raw_data=False
...
+        The dataframe analysed.
+
+    Returns
+    -------
+    matplotlib chart
+        A matrix plot of the variables containing the residuals from the respective partial regressions.
+    """
+
+    if len(x) == 2:
```
Comment (krajcsi) [inline cogstat/cogstat_chart.py:805]: I'm not sure what this function does. Seemingly, x is not a string. Please, revise the docstring, comments, and variable names

File: cogstat/cogstat_stat.py [hunk 7]
```diff
@@ -699,6 +699,168 @@ def variable_pair_standard_effect_size(data, meas_lev, sample=True, normality=No
                                                                                  classes="table_cs_pd"))
     return standardized_effect_size_result
 
+def calc_se_r2(r2, k, n):
```
Comment (krajcsi) [inline cogstat/cogstat_stat.py:702]: this function should go to the cogstat_stat_num module

File: cogstat/cogstat_stat.py [hunk 8]
```diff
@@ -699,6 +699,168 @@ def variable_pair_standard_effect_size(data, meas_lev, sample=True, normality=No
...
+    -------
+    float
+    """
+
+    numerator = 4*r2*((1-r2)**2)*((n-k-1)**2)
+    denominator = ((n**2) - 1)*(3+n)
+    se_r2 = (numerator/denominator)**0.5
+    return se_r2
+
+def calc_r2_ci(r2, alpha, n, k):
```
Comment (krajcsi) [inline cogstat/cogstat_stat.py:728]: this function should go to the cogstat_stat_num module

k shoud be before n?

shoud we set default alpha?

File: cogstat/cogstat_stat.py [hunk 9]
```diff
@@ -699,6 +699,168 @@ def variable_pair_standard_effect_size(data, meas_lev, sample=True, normality=No
...
+    Based on: Cohen, J., Cohen, P., West, S.G., and Aiken, L.S. (2003). Applied Multiple Regression/Correlation
+    Analysis for the Behavioral Sciences (3rd edition). Mahwah, NJ: Lawrence Earlbaum Associates. pp. 88
+    See also: https://stats.stackexchange.com/questions/175026/formula-for-95-confidence-interval-for-r2
+
+    Returns
+    -------
+    float
+    """
+
+    numerator = 4*r2*((1-r2)**2)*((n-k-1)**2)
```
Comment (krajcsi) [inline cogstat/cogstat_stat.py:723]: please, use spaces around operators (in the following lines too)

File: cogstat/cogstat_stat.py [hunk 10]
```diff
@@ -699,6 +699,168 @@ def variable_pair_standard_effect_size(data, meas_lev, sample=True, normality=No
...
+    See also: https://www.danielsoper.com/statcalc/formulas.aspx?id=28
+    Validated against: https://www.danielsoper.com/statcalc/calculator.aspx?id=28 on 02/06/2022 06:03
+
+    Returns
+    -------
+    list
+    """
+
+    se_r2 = calc_se_r2(r2,k,n)
+    t = stats.t.ppf((1-alpha)/2, n-k-1)
```
Comment (krajcsi) [inline cogstat/cogstat_stat.py:752]: use spaces around operators

File: cogstat/cogstat_stat.py [hunk 11]
```diff
@@ -699,6 +699,168 @@ def variable_pair_standard_effect_size(data, meas_lev, sample=True, normality=No
...
+        Name of the explanatory variables.
+    y : list of str
+        Name of the dependent variable.
+    normality: bool or None
+        True if variables follow a multivariate normal distribution, False otherwise. None if normality couldn't be
+        calculated or if the parameter was not specified.
+    homoscedasticity : bool or None
+        True if variables are homoscedastic, False otherwise. None if homoscedasticity couldn't be calculated or
+        if the parameter was not specified.
+    multicollinearity:
```
Comment (krajcsi) [inline cogstat/cogstat_stat.py:774]:
```suggestion
    multicollinearity : bool or None
```

File: cogstat/cogstat_stat.py [hunk 12]
```diff
@@ -699,6 +699,168 @@ def variable_pair_standard_effect_size(data, meas_lev, sample=True, normality=No
...
+        True for sample descriptives, False for population estimations.
+    result : statsmodels regression result object
+        The result of the multiple regression analyses.
+
+    Returns
+    -------
+    html text
+    """
+    # TODO validate
+    import pingouin as pg
```
Comment (krajcsi) [inline cogstat/cogstat_stat.py:786]: it's already immported

please, do not abbreviate

File: cogstat/cogstat_stat.py [hunk 13]
```diff
@@ -699,6 +699,168 @@ def variable_pair_standard_effect_size(data, meas_lev, sample=True, normality=No
...
+
+    standardized_effect_size_result = '<cs_h3>' + _('Standardized effect size') + '</cs_h3>' + "\n"
+
+    # Warnings based on the results of the assumption tests
+    if normality is None:
+        standardized_effect_size_result += '\n' + '<decision>' + _('Normality could not be calculated.') + ' ' + \
+                                           _('CIs may be biased.') + '</decision>'
+    elif not normality:
+        standardized_effect_size_result += '\n' + '<decision>' + \
+                                           _('Assumption of normality violated for CI calculations.') + ' ' + \
```
Comment (krajcsi) [inline cogstat/cogstat_stat.py:796]: do we need "for CI calculations"? consider this for the next lines too

File: cogstat/cogstat_stat.py [hunk 14]
```diff
@@ -699,6 +699,168 @@ def variable_pair_standard_effect_size(data, meas_lev, sample=True, normality=No
...
+    else:
+        standardized_effect_size_result += '\n' + '<decision>' + _('Assumption of multicollinearity for parameter '\
+                                                                   'and CI calculations met.') + '</decision>'
+
+    # Calculate effect sizes for sample or population
+    if sample:
+        pdf_result_corr = pd.DataFrame()
+        if result:
+            standardized_effect_size_result += "\n" + _('<i>R<sup>2</sup></i> = %0.3f' % result.rsquared) + "\n"
+    else:
```
Comment (krajcsi) [inline cogstat/cogstat_stat.py:829]:
```suggestion
    else:  # population
```

File: cogstat/cogstat_stat.py [hunk 15]
```diff
@@ -699,6 +699,168 @@ def variable_pair_standard_effect_size(data, meas_lev, sample=True, normality=No
...
+                                           + _('Assumption of multicollinearity violated for CI calculations.') + ' ' + \
+                                           _('CIs may be biased.') + '</decision>'
+    else:
+        standardized_effect_size_result += '\n' + '<decision>' + _('Assumption of multicollinearity for parameter '\
+                                                                   'and CI calculations met.') + '</decision>'
+
+    # Calculate effect sizes for sample or population
+    if sample:
+        pdf_result_corr = pd.DataFrame()
+        if result:
```
Comment (krajcsi) [inline cogstat/cogstat_stat.py:827]: do we need to check this? if result is not available, the whole function seems to be pointless (is there any other details it calculates?) And in some other lines it is not checked, which makes the solution incoherent, and pointless again. Please, double check this in the rest of the function too.

**Review by krajcsi — CHANGES_REQUESTED** [review]

File: cogstat/cogstat_chart.py [hunk 16]
```diff
@@ -725,6 +725,107 @@ def create_variable_pair_chart(data, meas_lev, x, y, result=None, raw_data=False
...
 
+def create_multiple_variable_chart(data, meas_lev, x, type="raw"):
+    """Draw a chart relating more than two variables displaying raw data
+
+    Parameters
+    ----------
+    data : pandas dataframe
+    meas_lev : {'int', 'ord', 'nom', 'unk'}
+        Measurement level of the variables
+    x : str
```
[the remark this answers is not in the document]
Comment (TomiSzucs) [reply cogstat/cogstat_chart.py]: Variable renamed to var_names.

File: cogstat/cogstat_chart.py [hunk 17]
```diff
@@ -725,6 +725,107 @@ def create_variable_pair_chart(data, meas_lev, x, y, result=None, raw_data=False
...
+        Displays raw data when 'raw', displays scatterplot matrix of explanatory variables when 'collinearity'.
+
+    Returns
+    -------
+    matplotlib chart
+        A matrix plot of the variables optionally containing the raw data.
+    """
+
+    if meas_lev == "int":
+        if type == "collinearity":
```
[the remark this answers is not in the document]
Comment (TomiSzucs) [reply cogstat/cogstat_chart.py]: True, modified.

File: cogstat/cogstat_chart.py (hunk 3 above)
Comment (TomiSzucs) [reply cogstat/cogstat_chart.py]: Function split into two separate funcitons.

File: cogstat/cogstat_chart.py (hunk 6 above)
Comment (TomiSzucs) [reply cogstat/cogstat_chart.py]: Docstring extended and comments added. I hope it's better explained now.

File: cogstat/cogstat_stat.py (hunk 7 above)
Comment (TomiSzucs) [reply cogstat/cogstat_stat.py]: Functions moved.

File: cogstat/cogstat_stat.py (hunk 8 above)
Comment (TomiSzucs) [reply cogstat/cogstat_stat.py]: Function moved, k moved and alpha set to 0.95 by default.

File: cogstat/cogstat_stat.py (hunk 13 above)
Comment (TomiSzucs) [reply cogstat/cogstat_stat.py]: After some thought, it is probably redundant, so I removed them.

File: cogstat/cogstat_chart.py (hunk 5 above)
Comment (TomiSzucs) [reply cogstat/cogstat_chart.py:810]: "data" is now in the first place.

File: cogstat/cogstat_chart.py (hunk 4 above)
Comment (TomiSzucs) [reply cogstat/cogstat_chart.py]: "x" renamed, docstring updated.

File: cogstat/cogstat_stat.py (hunk 12 above)
Comment (TomiSzucs) [reply cogstat/cogstat_stat.py]: True, it missed my attention. Fixed.

File: cogstat/cogstat_stat.py (hunk 10 above)
Comment (TomiSzucs) [reply cogstat/cogstat_stat.py]: Spaces added, it is much more readable now.

File: cogstat/cogstat_stat.py (hunk 9 above)
Comment (TomiSzucs) [reply cogstat/cogstat_stat.py]: Spaces added.

**Comment by krajcsi** [comment]
Looks good, thank you! We'll fix a few things later.

**Pull request merged**